### ISSUE TITLE
ThreadPerDriverTaskExecutor causes workes to stuck

### ISSUE THREAD

**Issue by davseitsev** [trinodb/trino#21512]
After Trino upgrade from version 409 to 444 we started wo experience issues with stuck workers. They refuse HTTP requests or the requests start to hang forever and desappear from discovery service.

Logs are full of messages like:
```
2024-04-09T13:31:03.821Z	WARN	node-state-poller-0	io.trino.metadata.RemoteNodeState	Node state update request to http://10.42.110.232:8181/v1/info/state has not returned in 35.08s
...
2024-04-09T13:33:38.872Z	ERROR	http-worker-529	io.trino.server.InternalAuthenticationManager	Internal authentication failed
io.jsonwebtoken.ExpiredJwtException: JWT expired 12872 milliseconds ago at 2024-04-09T13:33:26.000Z. Current time: 2024-04-09T13:33:38.872Z. Allowed clock skew: 0 milliseconds.
	at io.jsonwebtoken.impl.DefaultJwtParser.parse(DefaultJwtParser.java:682)
```
Where `10.42.110.232` is an IP of problematic worker.

Number of open file descriptors on the problematic worker dramatically increases:
![image](https://github.com/trinodb/trino/assets/1793410/666a0ee8-d78f-4060-b779-78c6b1fba6ce)

**Sometimes** we reach open sockets limit and got a lot of exceptions in logs like this:
```
2024-04-09T09:37:55.827Z	WARN	http-worker-124-acceptor-0@4e826fd4-http@44e08a7a{HTTP/1.1, (http/1.1, h2c)}{0.0.0.0:8181}	org.eclipse.jetty.server.AbstractConnector	Accept Failure
java.io.IOException: Too many open files
	at java.base/sun.nio.ch.Net.accept(Native Method)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.implAccept(ServerSocketChannelImpl.java:433)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:399)
	at org.eclipse.jetty.server.ServerConnector.accept(ServerConnector.java:397)
	at org.eclipse.jetty.server.AbstractConnector$Acceptor.run(AbstractConnector.java:698)
	at org.eclipse.jetty.util.thread.MonitoredQueuedThreadPool$1.run(MonitoredQueuedThreadPool.java:73)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:971)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1201)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1156)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```
But it happens not always.
On Trino 409 we had limit for open file descriptors 128K on workes. When we met this issue we increased the limit to 256K, but it doesn't help. This limit is exceeded very quickly.

Number of threads also start to increase:
![image](https://github.com/trinodb/trino/assets/1793410/20265861-a805-45e2-a9d4-937273fb5fa7)

Thread dump shows that single random thread blocks many other threads.
Example 1:
![image](https://github.com/trinodb/trino/assets/1793410/1be3592a-b150-471d-af8d-da43f2d8bc32)

Blocking thread stack trace (Full thread dump [trino-worker-D0409-T0940-25683.tdump.txt](https://github.com/trinodb/trino/files/14943812/trino-worker-D0409-T0940-25683.tdump.txt)) :

```
Task-20240409_093501_07390_nd69d.3.16.0-5123
Stack Trace is:
java.lang.Thread.State: RUNNABLE
at java.lang.Thread.start0(java.base@21.0.2/Native Method)
at java.lang.Thread.start(java.base@21.0.2/Thread.java:1553)
- locked <0x00007f6d32039e00> (a java.lang.Thread)
at java.lang.System$2.start(java.base@21.0.2/System.java:2577)
at jdk.internal.vm.SharedThreadContainer.start(java.base@21.0.2/SharedThreadContainer.java:152)
at java.util.concurrent.ThreadPoolExecutor.addWorker(java.base@21.0.2/ThreadPoolExecutor.java:953)
at java.util.concurrent.ThreadPoolExecutor.execute(java.base@21.0.2/ThreadPoolExecutor.java:1375)
at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:640)
at java.util.concurrent.AbstractExecutorService.submit(java.base@21.0.2/AbstractExecutorService.java:134)
at com.google.common.util.concurrent.AbstractListeningExecutorService.submit(AbstractListeningExecutorService.java:73)
at io.trino.execution.executor.scheduler.FairScheduler.submit(FairScheduler.java:161)
- locked <0x00007f57374acb88> (a io.trino.execution.executor.scheduler.FairScheduler)
at io.trino.execution.executor.dedicated.TaskEntry.runSplit(TaskEntry.java:136)
- locked <0x00007f6d32000000> (a io.trino.execution.executor.dedicated.TaskEntry)
at io.trino.execution.executor.dedicated.ThreadPerDriverTaskExecutor.enqueueSplits(ThreadPerDriverTaskExecutor.java:157)
- locked <0x00007f57373c35b0> (a io.trino.execution.executor.dedicated.ThreadPerDriverTaskExecutor)
at io.trino.execution.SqlTaskExecution.enqueueDriverSplitRunner(SqlTaskExecution.java:400)
- locked <0x00007f6285ff4580> (a io.trino.execution.SqlTaskExecution)
at io.trino.execution.SqlTaskExecution.scheduleDriversForTaskLifeCycle(SqlTaskExecution.java:389)
at io.trino.execution.SqlTaskExecution.start(SqlTaskExecution.java:205)
- locked <0x00007f6285ff4580> (a io.trino.execution.SqlTaskExecution)
at io.trino.execution.SqlTask.tryCreateSqlTaskExecution(SqlTask.java:561)
- locked <0x00007f5eb809b388> (a java.lang.Object)
at io.trino.execution.SqlTask.updateTask(SqlTask.java:503)
at io.trino.execution.SqlTaskManager.doUpdateTask(SqlTaskManager.java:555)
at io.trino.execution.SqlTaskManager.lambda$updateTask$9(SqlTaskManager.java:489)
at io.trino.execution.SqlTaskManager$$Lambda/0x00007f566ab68000.call(Unknown Source)
at io.trino.$gen.Trino_444____20240409_090553_2.call(Unknown Source)
at io.trino.execution.SqlTaskManager.updateTask(SqlTaskManager.java:489)
at io.trino.server.TaskResource.createOrUpdateTask(TaskResource.java:153)
...
```

Example 2:
![image](https://github.com/trinodb/trino/assets/1793410/61d8ac6e-9691-46c5-a9c1-2c8fd406e46b)

Blocking thread stack trace (Full thread dump [trino-worker-D0409-T0949-25683.tdump.txt](https://github.com/trinodb/trino/files/14943841/trino-worker-D0409-T0949-25683.tdump.txt)):
```
SplitRunner-24734
Stack Trace is:
java.lang.Thread.State: RUNNABLE
at java.lang.Thread.start0(java.base@21.0.2/Native Method)
at java.lang.Thread.start(java.base@21.0.2/Thread.java:1553)
- locked <0x00007f6db5a217a8> (a java.lang.Thread)
at java.lang.System$2.start(java.base@21.0.2/System.java:2577)
at jdk.internal.vm.SharedThreadContainer.start(java.base@21.0.2/SharedThreadContainer.java:152)
at java.util.concurrent.ThreadPoolExecutor.addWorker(java.base@21.0.2/ThreadPoolExecutor.java:953)
at java.util.concurrent.ThreadPoolExecutor.execute(java.base@21.0.2/ThreadPoolExecutor.java:1375)
at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:640)
at java.util.concurrent.AbstractExecutorService.submit(java.base@21.0.2/AbstractExecutorService.java:134)
at com.google.common.util.concurrent.AbstractListeningExecutorService.submit(AbstractListeningExecutorService.java:73)
at io.trino.execution.executor.scheduler.FairScheduler.submit(FairScheduler.java:161)
- locked <0x00007f57374acb88> (a io.trino.execution.executor.scheduler.FairScheduler)
at io.trino.execution.executor.dedicated.TaskEntry.runSplit(TaskEntry.java:136)
- locked <0x00007f5cbd1881d8> (a io.trino.execution.executor.dedicated.TaskEntry)
at io.trino.execution.executor.dedicated.TaskEntry.dequeueAndRunLeafSplit(TaskEntry.java:116)
- locked <0x00007f5cbd1881d8> (a io.trino.execution.executor.dedicated.TaskEntry)
at io.trino.execution.executor.dedicated.ThreadPerDriverTaskExecutor.scheduleLeafSplit(ThreadPerDriverTaskExecutor.java:170)
at io.trino.execution.executor.dedicated.ThreadPerDriverTaskExecutor.scheduleMoreLeafSplits(ThreadPerDriverTaskExecutor.java:190)
- locked <0x00007f57373c35b0> (a io.trino.execution.executor.dedicated.ThreadPerDriverTaskExecutor)
at io.trino.execution.executor.dedicated.ThreadPerDriverTaskExecutor.leafSplitDone(ThreadPerDriverTaskExecutor.java:181)
- locked <0x00007f57373c35b0> (a io.trino.execution.executor.dedicated.ThreadPerDriverTaskExecutor)
at io.trino.execution.executor.dedicated.ThreadPerDriverTaskExecutor$$Lambda/0x00007f566abdfd50.run(Unknown Source)
at io.trino.execution.executor.dedicated.TaskEntry.lambda$dequeueAndRunLeafSplit$0(TaskEntry.java:119)
at io.trino.execution.executor.dedicated.TaskEntry$$Lambda/0x00007f566abe2ad0.run(Unknown Source)
at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
...
```

We have changed value of `experimental.thread-per-driver-scheduler-enabled` to `false` and now the issue is not reproduced, our cluster looks stable
https://github.com/trinodb/trino/blob/b7a161a422850c35b992f3ba26ac1a2f8bc9eb54/core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java#L110-L115
I will update if anything changes.

Let me know if I can provide more information.

**Comment by raunaqmorarka** [comment]
cc: @martint @findepi

**Comment by findepi** [comment]
@davseitsev appreciate detailed analysis, that's very helpful!

**Comment by nesat** [comment]
Thanks a lot for sharing your findings and a potential solution for the problem. We had a similar issue on our Trino cluster started after we upgraded Trino from version 426 to 442. Workers started to have exchange issues more frequently with the logs:
```
2024-04-25T10:24:55.409Z        WARN    node-state-poller-0     io.trino.metadata.RemoteNodeState       Node state update request to https://xxx.ip:8081/v1/info/state has not returned in 10.02s
2024-04-25T10:24:55.410Z        WARN    node-state-poller-0     io.trino.metadata.RemoteNodeState       Node state update request to https://xxx.ip:8081/v1/info/state has not returned in 10.02s
2024-04-25T10:24:58.184Z        INFO    Thread-47       io.airlift.bootstrap.LifeCycleManager   JVM is shutting down, cleaning up
2024-04-25T10:24:58.185Z        INFO    Thread-42       io.airlift.bootstrap.LifeCycleManager   JVM is shutting down, cleaning up
```

I see the `experimental.thread-per-driver-scheduler-enabled` was enabled since version 438(https://github.com/trinodb/trino/pull/20451) so we decided to disable it and it helped with stabilizing the cluster. We still see some exchange issues but the number of impacted queries used to be 14x more before disabling the feature.

**Comment by Pluies** [comment]
Chiming in for more feedback - we're seeing exactly the same symptoms here on an update from 436 to 444. Nodes are going down, queries failing left and right, and a very very unstable cluster altogether.

If it helps at all, we're also seeing a ~30% query speed slowdown on smaller benchmarks, before nodes go unresponsive. Running the benchmark for longer (not very long in the grand scheme of things, maybe 5 to 10 minutes at ~half our current prod concurrency) causes unresponsive nodes.

I can also confirm setting `experimental.thread-per-driver-scheduler-enabled=false` fixes it completely: performance is restored to Trino 436 levels and nodes are stable.

**Comment by martint** [comment]
Are you able to narrow it down to specific query shapes? It would help immensely to debug if we can get some examples that cause the cluster to fall over, or that perform more slowly.

**Comment by Pluies** [comment]
@martint let me dig into our benchmark and see if I can narrow it down to specific shapes 👍

**Comment by Pluies** [comment]
@martint well it's been a bit of a wild goose chase. I have a list of 100 queries that do cause the issue, but I've been unable to pinpoint the issue to any single query in particular. Happy to share this list in private if it helps.

All these queries:
- Are read-only
- Use the Delta Lake connector
- Use S3 caching

@davseitsev are there any communalities with your workloads? 🤔 

I can't imagine this issue affects every Trino workloads or there would have been much more noise about it after the 436 release.

**Comment by martint** [comment]
Do the queries have joins, union, many stages? Does it happen if you load the cluster with any specific type of query, or only when you mix them up?

**Comment by okayhooni** [comment]
Thanks a lot. You saved my life :)

**Comment by sopel39** [comment]
@Pluies is the issue fixed or is still present?

**Comment by Pluies** [comment]
@sopel39 I haven't tested recently. Which version would you expect a fix to be present in?

**Comment by sopel39** [comment]
@martint ^^?

**Comment by martint** [comment]
The issue is still present. We need to be able to reproduce it and figure out the root cause.

**Comment by priyanshu-delhivery** [comment]
@Pluies I observed the same behaviour when running a certain set of queries, but I could not reproduce it when running them 1 by 1. This must be some concurrency-related thing. One thing which I observed is that some of my queries in the query set (4 out of 80) fail with the error: operator not found bigint = bigint, before the cluster starts underperforming and the worker get unresponsive. Also, these queries ran successfully when triggered 1 by 1.

**Comment by findepi** [comment]
> fail with the error: operator not found bigint = bigint

there used to be a bug in JVM's JIT that was causing spurious errors like that (https://github.com/trinodb/trino/issues/5758), but it's believed to have been fixed.

i don't see how an error like that could be related to ThreadPerDriverTaskExecutor, so likely it is a separate problem.
Do you have full stacktrace?

**Comment by hqbhoho** [comment]
is the issue fixed or is still present?

**Comment by netanelm-upstream** [comment]
I tried to disable `experimental.thread-per-driver-scheduler-enabled ` and still have the same symptoms.
Using Trino 442 several months, but just last week encountered this issue.
Any ideas what else to check? (more details in the [slack channel](https://trinodb.slack.com/archives/CGB0QHWSW/p1723705104766299) )

**Comment by martint** [comment]
Added diagnostics to help track down the issue: https://github.com/trinodb/trino/pull/23045

**Comment by jonathanmyers** [comment]
We have been dealing with stuck workers on our clusters that run merge and maintenance queries against iceberg.  Typically a given worker will fail a health check multiple times over a 20 or 30 minute period.  To give a sense of the frequency, we run 2 of the same merge query across ~2000 tables every hour.  

As with others in the thread, `experimental.thread-per-driver-scheduler-enabled=false` has made the issue disappear entirely.  Happy to provide more info if it would help get to the bottom of this problem.

**Comment by losipiuk** [comment]
@jonathanmyers thanks for more info. How many queries do you run in parallel?

**Comment by lozbrown** [comment]
> We have been dealing with stuck workers on our clusters that run merge and maintenance queries against iceberg. Typically a given worker will fail a health check multiple times over a 20 or 30 minute period. To give a sense of the frequency, we run 2 of the same merge query across ~2000 tables every hour.
> 
> As with others in the thread, `experimental.thread-per-driver-scheduler-enabled=false` has made the issue disappear entirely. Happy to provide more info if it would help get to the bottom of this problem.

what version are you on, can you provide logs from the workers

in 455 there is the change @martint mentioned that logs additional info about thread state

**Comment by jonathanmyers** [comment]
We're on trino 458, and we send 50 merge queries at a time, 100 for our iceberg optimize tasks. We run a baseline of 8 m7g.8xlarge workers and scale up to 16 of those for table maintenance.  I've enabled the debug logs and re-enabled thread per driver in our test environment, should be able to get those tomorrow.  If there's anything specific from those logs we want let me know, otherwise I'll try to get a sample from around the time a node starts hanging.  In the past there has not been anything interesting in the logs-- the rest of the cluster throws errors related to it going down but that's it. Hopefully the debug logs will shed some light.

**Comment by losipiuk** [comment]
> If there's anything specific from those logs we want let me know, otherwise I'll try to get a sample from around the time a node starts hanging.

It should not be too much volume. If you enable DEBUG log level on workers for `io.trino.execution.executor.dedicated.ThreadPerDriverTaskExecutor` when cluster is unstable it should log a diagnostic snippet every 30s.
You can do that at runtime via JMX.

**Comment by jonathanmyers** [comment]
Here are the debug logs from a stuck worker [stuck_worker_logs.txt](https://github.com/user-attachments/files/17999137/stuck_worker_logs.txt)

**Comment by maswin** [comment]
Internally after upgrading to Trino-440 we faced the same issue. One thing is, this happened only in clusters that had lot of writes happening. Read-only clusters didn't show up this issue.

**Comment by lozbrown** [comment]
Is there any clear update here. should this continue to be disabled?

**Comment by losipiuk** [comment]
@lukasz-stec connect me if I am wrong but IIRC the current state of understanding is that we just get into trouble when we are getting into situation where worker has planty of Drivers runnable which translates to plenty of Threads. As with java 23 those threads are OS threads, after crossing some threshold we just make JVM unstable and everything falls apart. 
If you have workflow which can put worker in such state then `ThreadPerDriverTaskExecutor` is not for you and you should turn it off. Situation may (should) improve with Java 24 when virtual threads are supposed to be usable.

cc: @wendigo

**Comment by lukasz-stec** [comment]
There has also been some progress in resolving this issue.
We can see this reproduced by running multiple, like 8, concurrent tpcds q04 queries. This is because q04 has a lot of intermediate splits that end up all running on a dedicated thread.

There was an effort by @hashhar to fix the issue with virtual threads but since the JVM 23 still blocks carrier threads on `synchronized` it was not working well. One of the options is to wait for the improvements in virtual threads, another one is to replace `synchronized` with `java.util.concurrent.locks`.

**Comment by wendigo** [comment]
@lukasz-stec this will pin either way in guava's loadingcache

**Comment by martint** [comment]
@losipiuk, there's something else going on. It's not just having too many drivers. There's some condition that causes the system to lock up. 

The data point @maswin provided is interesting. We should follow up on that.

**Comment by lozbrown** [comment]
> The data point [@maswin](https://github.com/maswin) provided is interesting. We should follow up on that.

We have experienced this issue and don't do any writing with trino

**Comment by lozbrown** [comment]
Could someone explain what the advantage of ThreadPerDriver executor is that we miss out on by disabling it?

**Comment by churromorales** [comment]
should this config be on by default?

**Comment by dc-orz** [comment]
I have encountered similar issues where the performance of ThreadPerDriverTaskExecutor is poor under high QPS scenarios (e.g., throughput stress testing). From my understanding, when there are many drivers on a worker, using ThreadPerDriverTaskExecutor might not be suitable. After switching to TimeSharingTaskExecutor, the throughput performance improved significantly.

**Comment by mayurjpatel** [comment]
Is `ThreadPerDriverTaskExecutor` meant to use virtual threads or platform threads? [FairScheduler.java](https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/execution/executor/scheduler/FairScheduler.java#L80) is using platform threads for tasks executor and I think this could explain the poor performance for `ThreadPerDriverTaskExecutor` under high QPS clusters compared with `TimeSharingTaskExecutor`.

**Comment by hashhar** [comment]
yes it is using platform threads for now. But I did try with virtual threads too and it helps but there is still a gap. Also this issue is more about a bug than a performance issue.

See https://github.com/trinodb/trino/compare/master...hashhar:trino:hashhar/virtual-sched-threads for the changes I did when I tried with Loom preview.

**Comment by dc-orz** [comment]
> yes it is using platform threads for now. But I did try with virtual threads too and it helps but there is still a gap. Also this issue is more about a bug than a performance issue.
> 
> See [master...hashhar:trino:hashhar/virtual-sched-threads](https://github.com/trinodb/trino/compare/master...hashhar:trino:hashhar/virtual-sched-threads) for the changes I did when I tried with Loom preview.

@hashhar What is your maxConcurrentTasks configuration?
I think this value should be set as large as possible when using virtual threads. This effectively delegates scheduling to the JDK's virtual thread scheduler. I’m not sure how its efficiency compares to Trino’s MLFQ implementation.

**Comment by wendigo** [comment]
Does it still happen on 478 folks? :)

**Comment by leonaascimento** [comment]
I can confirm this still happens on Trino 479.

We recently changed our deployment configuration and accidentally dropped `experimental.thread-per-driver-scheduler-enabled=false`. As soon as the new config rolled out to the cluster serving ad-hoc queries to our business users, PAGE_TRANSPORT_TIMEOUT and PAGE_TRANSPORT_ERROR failures spiked. Re-adding the setting resolved the issue.

Comparing the failed queries against successful queries from the period, the failures are heavily skewed toward complex (actually, poorly written) queries:

|                         | Failed        | Successful     |
|-------------------------|---------------|----------------|
| Avg stages              | 20.8          | 5.0            |
| Avg tasks               | 190 (p95 715) | 35.7 (p95 169) |
| Avg JOINs               | 6.25          | 1.18           |
| ≥5 joins                | 49%           | 9%             |
| CTEs                    | 60%           | 15%            |
| Window functions        | 37%           | 7.5%           |
| UNION                   | 15%           | 3.7%           |
| Splits p50              | 4,096         | 37             |
| Queries hitting full GC | 5.4%          | 1.0%           |

In short, queries with many stages/tasks/splits were disproportionately affected.

**Comment by findepi** [comment]
@martint see above

**Comment by martint** [comment]
@leonaascimento, I posted a potential fix (#30842) for the issue. Would you be able to test it with your workloads?

**Comment by leonaascimento** [comment]
Sure @martint, probably next week.

**Comment by wendigo** [comment]
@leonaascimento any news?

**Comment by martint** [comment]
Reopening until fix is validated
